src/jobserver-posix.cc

```diff
@@ -0,0 +1,71 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "jobserver.h"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <cassert>
+#include <cstring>
+
+#include "util.h"
+
+void Jobserver::Init() {
+  assert(fd_ < 0);
+
+  if (!ParseJobserverAuth("fifo")) {
+    return;
+  }
+
+  const char* jobserver = jobserver_name_.c_str();
+
+  fd_ = open(jobserver, O_NONBLOCK | O_CLOEXEC | O_RDWR);
+  if (fd_ < 0) {
+    Fatal("failed to open jobserver: %s: %s", jobserver, strerror(errno));
+  }
+
+  Info("using jobserver: %s", jobserver);
+}
+
+Jobserver::~Jobserver() {
+  assert(token_count_ == 0);
+
+  if (fd_ >= 0) {
+    close(fd_);
+  }
+}
+
+bool Jobserver::Enabled() const {
+  return fd_ >= 0;
+}
+
+bool Jobserver::AcquireToken() {
+  char token;
+  int res = read(fd_, &token, 1);
+  if (res < 0 && errno != EAGAIN && errno != EWOULDBLOCK) {
+    Fatal("failed to read jobserver token: %s", strerror(errno));
+  }
+
+  return res > 0;
+}
+
+void Jobserver::ReleaseToken() {
+  char token = '+';
```

Comment:
from the Make manual:

> It’s important that when you release the job slot, you write back the same character you read. Don’t assume that all tokens are the same character; different characters may have different meanings to GNU make.

I have not looked at Make's code relevant to this yet, so I don't know the significance. It might be that the character being anything other than a `+` is a rare occurence... maybe someone else can clarify how important it is.

I will follow up with another comment with more information as I continue to play with the PR's changes and review Make's source.

Reply:
Yeah, I've noticed the same from the Make manual. It might make sense to store the acquired token in the (recently added) `Edge::acquired_job_server_token_` member. I'll look into it.

Reply:
After looking at GNU Make source history, I changed my mind. I don't think this is important. There is no sign that they are going to introduce new characters that have special meaning, and if they somehow do that, it would be pretty easy to add support for it later. Let's save the idea for the future...